compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/codegen/SwitchGenerator.kt

```diff
@@ -0,0 +1,181 @@
+import org.jetbrains.kotlin.codegen.StackValue.*
+import org.jetbrains.kotlin.ir.IrElement
+import org.jetbrains.kotlin.ir.expressions.*
+import org.jetbrains.kotlin.ir.types.*
+import org.jetbrains.kotlin.ir.util.isTrueConst
+import org.jetbrains.kotlin.types.KotlinType
+import org.jetbrains.org.objectweb.asm.Label
+import org.jetbrains.org.objectweb.asm.Type
+import java.util.*
+
+// TODO: Char, Byte, Short, String, enum, eliminate the temporary variable
+class SwitchGenerator(private val expression: IrWhen, private val data: BlockInfo, private val codegen: ExpressionCodegen) {
+    private val endLabel = Label()
+    private var defaultLabel = endLabel
+    private lateinit var elseExpression: IrExpression
+    private val cases = ArrayList<Pair<Int, Label>>()
+    private val thenExpressions = ArrayList<Pair<IrExpression, Label>>()
+    private var subject: IrGetValue? = null
+
+    private val mv = codegen.mv
+
+    // @return true if the IrWhen can be emitted as lookupswitch or tableswitch.
+    fun match(): Boolean {
+        for (branch in expression.branches) {
+            if (branch is IrElseBranch) {
+                elseExpression = branch.result
+                defaultLabel = Label()
+            } else {
+                val thenLabel = Label()
+                thenExpressions.add(Pair(branch.result, thenLabel))
+                if (!matchConditions(branch.condition, thenLabel))
+                    return false
+            }
+        }
+
+        // IF is more compact when there are only 1 or fewer branches, in addition to else.
+        if (cases.size <= 1)
+            return false
+
+        return true
+    }
+
+    // psi2ir lowers multiple cases to nested conditions. For example,
+    //
+    // when (subject) {
+    //   a, b, c -> action
+    // }
+    //
+    // is lowered to
+    //
+    // if (if (subject == a)
+    //       true
+    //     else
+    //       if (subject == b)
+    //         true
+    //       else
+    //         subject == c) {
+    //     action
+    // }
+    //
+    // @return true if the conditions are equality checks of constants.
+    private fun matchConditions(condition: IrExpression, thenLabel: Label): Boolean {
+        if (condition is IrCall) {
+            // Leaf nodes: a lookup/table switch can be used if:
+            // 1. All branches are CALL 'EQEQ(Any?, Any?)': Boolean
+            // 2. All types of variables involved in comparison are in the same group of Char/Byte/Short/Int, String or enum.
+            // 3. All arg0 refer to the same value.
+            // 4. All arg1 are IrConst<*>.
+
+            if (condition.symbol != codegen.classCodegen.context.irBuiltIns.eqeqSymbol)
+                return false
+
+            // TODO: The check relies on the implementation that subject is always lowered to a local variable.
+            val candidate = condition.getValueArgument(0) as? IrGetValue ?: return false
+            subject = subject ?: candidate
```

Comment:
Please consider rewriting this to avoid modification of mutable fields, as it isn't very easy to debug code operating with mutable state. For example, instead of modifying `subject` and `cases` along the way of computing whether this is an optimizable when-condition, consider making `matchConditions` return a nullable list of case + 
subject, where null would mean that this is not an optimizable when-condition. Something like this:
```
private data class ConditionInfo(val case: Pair<Int, Label>, val subject: IrGetValue)

...

private fun matchConditions(condition: IrExpression, thenLabel: Label): List<ConditionInfo>? { ... }
```

At the call site, we can then return if any `matchConditions` returned null, ensure that all `subject`s of these conditions are the same, and flatten the results to pass to `gen`. An example advantage here is that if I'm debugging this code I can easily go backwards and forwards, re-evaluate `matchConditions` on anything I like without side effects, etc.

Reply:
Done. matchConditions now only get a list of IrCall without having side effects. The result can also be reused, e.g., for String and Enum.

Reply:
Thanks!